src/meta/meta_bulk_load_service.cpp

```diff
@@ -1349,9 +1349,197 @@ void bulk_load_service::do_sync_partition(const gpid &pid, std::string &partitio
+        return false;
+    }
+
+    // partition_bulk_load_info_size should not be greater than partition_count
+    if (partition_count < pinfo_size) {
+        derror_f("app({}) has invalid count, app partition_count = {}, remote "
+                 "partition_bulk_load_info count = {}",
+                 ainfo.app_name,
+                 partition_count,
+                 pinfo_size);
+        return false;
+    }
+
+    // partition_bulk_load_info_size is not equal to partition_count can only be happended when app
+    // status is downloading, consider the following condition:
+    // when starting bulk load, meta server will create app_bulk_load_dir and
+    // partition_bulk_load_dir on remote storage
+    // however, meta server crash when create app directory and part of partition directory
+    // when meta server recover, partition directory count is less than partition_count
+    if (pinfo_size != partition_count && ainfo.status != bulk_load_status::BLS_DOWNLOADING) {
+        derror_f("app({}) bulk_load_status = {}, but there are {} partitions lack "
+                 "partition_bulk_load dir",
+                 ainfo.app_name,
+                 dsn::enum_to_string(ainfo.status),
+                 partition_count - pinfo_size);
+        return false;
+    }
+
+    return true;
+}
+
+// ThreadPool: THREAD_POOL_META_SERVER
+/*static*/ bool bulk_load_service::validate_partition(
+    const app_bulk_load_info &ainfo,
+    const std::unordered_map<int32_t, partition_bulk_load_info> &pinfo_map,
+    const int32_t different_status_count)
+{
+    const auto app_status = ainfo.status;
+    bool is_valid = true;
+
+    switch (app_status) {
+    case bulk_load_status::BLS_DOWNLOADING:
+        // if app status is downloading, partition status has no limit, because when bulk load meet
+        // recoverable errors, will rollback to downloading
+        // partition directory count is allowed less than partition_count, but it is impossible with
+        // some partition bulk load status is not downloading and some partition directroy is
+        // missing on remote storage
+        if (ainfo.partition_count - pinfo_map.size() > 0 && different_status_count > 0) {
+            derror_f("app({}) bulk_load_status = {}, there are {} partitions status is different "
+                     "from app, and {} partitions not existed, this is invalid",
+                     ainfo.app_name,
+                     dsn::enum_to_string(app_status),
+                     different_status_count,
+                     ainfo.partition_count - pinfo_map.size());
+            is_valid = false;
+        }
+        break;
+    case bulk_load_status::BLS_DOWNLOADED:
+    case bulk_load_status::BLS_INGESTING: {
+        // if app status is downloaded, valid partition status is downloaded or ingesting
+        // if app status is ingesting, valid partition status is ingesting or succeed
+        const auto valid_status = (app_status == bulk_load_status::BLS_DOWNLOADED)
```

Comment:
The same status with `app_status ` is valid too, how about rename `valid_status ` to `other_ valid_status `?